linkerd/duplex/src/lib.rs

```diff
@@ -85,81 +97,170 @@ where
+    /// the number of bytes read. If the buffer already has data, no new data
+    /// will be read.
+    fn poll_buffer(&mut self, cx: &mut Context<'_>) -> io::Poll<Buffered> {
+        // Buffer data only if no data is buffered.
+        //
+        // TODO should we read more data as long as there's buffer capacity?
+        // To do this, we'd have to get more complex about handling EOF.
+        if let Some(buf) = self.buf.as_mut() {
+            if buf.has_remaining() {
+                // Data was already buffered, so just return 0 immediately.
+                trace!(direction = %self.direction, remaining = buf.remaining(), "skipping read");
+                return Poll::Ready(Ok(Buffered::NotEmpty));
+            }
 
-                trace!(direction = %self.direction, "reading");
-                let n = ready!(io::poll_read_buf(Pin::new(&mut self.io), cx, buf))?;
-                trace!(direction = %self.direction, "read {}B", n);
+            buf.reset();
+            trace!(direction = %self.direction, "reading");
+            let sz = ready!(io::poll_read_buf(Pin::new(&mut self.io), cx, buf))?;
+            trace!(direction = %self.direction, "read {}B", sz);
 
-                is_eof = n == 0;
+            // If data was read, return the number of bytes read.
+            if sz > 0 {
+                return Poll::Ready(Ok(Buffered::Read(sz)));
             }
         }
-        if is_eof {
-            trace!("eof");
-            self.buf = None;
-        }
 
-        Poll::Ready(Ok(()))
+        // No more data can be read.
+        trace!("eof");
+        self.buf = None;
+        Poll::Ready(Ok(Buffered::Eof))
+    }
+
+    fn poll_flush<U: AsyncWrite + Unpin>(
+        &mut self,
+        dst: &mut HalfDuplex<U>,
+        cx: &mut Context<'_>,
+    ) -> io::Poll<()> {
+        trace!(direction = %self.direction, "flushing");
+        let poll = Pin::new(&mut dst.io).poll_flush(cx);
+        self.flushing = poll.is_pending();
+        if poll.is_ready() {
+            trace!(direction = %self.direction, "flushed");
+        }
+        poll
     }
 
-    fn poll_write_into<U>(
+    /// Writes as much buffered data as possible, returning the number of bytes written.
+    fn drain_into<U: AsyncWrite + Unpin>(
         &mut self,
         dst: &mut HalfDuplex<U>,
         cx: &mut Context<'_>,
-    ) -> Poll<Result<(), io::Error>>
-    where
-        U: AsyncWrite + Unpin,
-    {
-        if let Some(ref mut buf) = self.buf {
+    ) -> io::Result<Drained> {
+        let mut sz = 0;
+
+        if let Some(buf) = self.buf.as_mut() {
             while buf.has_remaining() {
                 trace!(direction = %self.direction, "writing {}B", buf.remaining());
-                let n = ready!(io::poll_write_buf(Pin::new(&mut dst.io), cx, buf))?;
+                let n = match io::poll_write_buf(Pin::new(&mut dst.io), cx, buf)? {
+                    Poll::Pending => return Ok(Drained::Partial(sz)),
```

Comment:
AFAICT `sz` is never used?

Reply:
similarly, not used but useful from the api pov.